# FINAL PROJECT 2 
# Prediksi Cuaca Harian di Australia
Anggota Kelompok 8:
- Rijal Muhammad Farizky
- Diva Nabila

## Project Overview
### Latar Belakang
`Cuaca` menjadi faktor penting yang dapat mempengaruhi kehidupan sehari-hari di berbagai sektor ekonomi, termasuk pertanian, pariwisata, dan manajemen bencana. Di Australia, cuaca yang ekstrem seperti kekeringan dan banjir dapat memiliki dampak yang signifikan pada kehidupan dan ekonomi masyarakat. Oleh karena itu, pengembangan sistem prediksi cuaca yang akurat berdasarkan data historis adalah suatu kebutuhan mendesak.

Dataset `Rain in Australia` berisi pengamatan cuaca harian selama sekitar 10 tahun dari berbagai stasiun cuaca di Australia. Dengan dataset ini kita dapat melakukan eksplorasi dan membangun model machine learning untuk prediksi cuaca dengan baik.

### Tujuan
1. `Exploratory Data Analysis` : Melakukan eksplorasi data yang mendalam untuk memahami pola cuaca di Australia selama 10 tahun terakhir, mengidentifikasi tren, variabilitas musiman, dan hubungan antar variabel cuaca.

2. `Prediksi Cuaca Harian`: Mengembangkan model prediksi cuaca harian menggunakan dua pendekatan: `Logistic Regression` dan `Support Vector Machine (SVM)`.

3. `Evaluasi Model`: Mengukur kinerja model menggunakan metrik evaluasi seperti akurasi, presisi, recall, dan F1-score. Melakukan perbandingan antara model Logistic Regression dan SVM untuk menentukan model mana yang memberikan prediksi terbaik.

4. `Implementasi dan Pengujian`: Melakukan deployment model Logistic Regression dan Support Vector Machine (SVM) ke dalam sistem yang dapat memberikan prediksi cuaca harian secara real-time berdasarkan data cuaca terbaru.

## Pustaka yang Digunakan

In [1]:
import warnings
warnings.filterwarnings('ignore')

%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import pickle

from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_regression
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, confusion_matrix, accuracy_score

## Data Loading
**Load Dataset**

In [5]:
df_wth = pd.read_csv('../dataset/weatherAUS.csv')

**Row and Column of Dataset**

In [6]:
df_wth.shape

(145460, 23)

**The Data**

In [40]:
df_wth

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0        Albury     13.4     22.9       0.6          4.8       8.4   
1        Albury      7.4     25.1       0.0          4.8       8.4   
2        Albury     12.9     25.7       0.0          4.8       8.4   
3        Albury      9.2     28.0       0.0          4.8       8.4   
4        Albury     17.5     32.3       1.0          4.8       8.4   
...         ...      ...      ...       ...          ...       ...   
145454    Uluru      3.5     21.8       0.0          4.8       8.4   
145455    Uluru      2.8     23.4       0.0          4.8       8.4   
145456    Uluru      3.6     25.3       0.0          4.8       8.4   
145457    Uluru      5.4     26.9       0.0          4.8       8.4   
145458    Uluru      7.8     27.0       0.0          4.8       8.4   

       WindGustDir  WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  \
0                W           44.0          W        WNW  ...         71.0   
1              WNW           44.0        NNW        WSW  ...         44.0   
2              WSW           46.0          W        WSW  ...         38.0   
3               NE           24.0         SE          E  ...         45.0   
4                W           41.0        ENE         NW  ...         82.0   
...            ...            ...        ...        ...  ...          ...   
145454           E           31.0        ESE          E  ...         59.0   
145455           E           31.0         SE        ENE  ...         51.0   
145456         NNW           22.0         SE          N  ...         56.0   
145457           N           37.0         SE        WNW  ...         53.0   
145458          SE           28.0        SSE          N  ...         51.0   

        Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
0              22.0       1007.7       1007.1       8.0       5.0     16.9   
1              25.0       1010.6       1007.8       5.0       5.0     17.2   
2              30.0       1007.6       1008.7       5.0       2.0     21.0   
3              16.0       1017.6       1012.8       5.0       5.0     18.1   
4              33.0       1010.8       1006.0       7.0       8.0     17.8   
...             ...          ...          ...       ...       ...      ...   
145454         27.0       1024.7       1021.2       5.0       5.0      9.4   
145455         24.0       1024.6       1020.3       5.0       5.0     10.1   
145456         21.0       1023.5       1019.1       5.0       5.0     10.9   
145457         24.0       1021.0       1016.8       5.0       5.0     12.5   
145458         24.0       1019.4       1016.5       3.0       2.0     15.1   

        Temp3pm  RainToday  RainTomorrow  
0          21.8         No            No  
1          24.3         No            No  
2          23.2         No            No  
3          26.5         No            No  
4          29.7         No            No  
...         ...        ...           ...  
145454     20.9         No            No  
145455     22.4         No            No  
145456     24.5         No            No  
145457     26.1         No            No  
145458     26.0         No            No  

[123710 rows x 22 columns]

**Informasi detail mengenai dataset**

In [41]:
df_wth.info()

<class 'pandas.core.frame.DataFrame'>
Index: 123710 entries, 0 to 145458
Data columns (total 22 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Location       123710 non-null  object 
 1   MinTemp        123710 non-null  float64
 2   MaxTemp        123710 non-null  float64
 3   Rainfall       123710 non-null  float64
 4   Evaporation    123710 non-null  float64
 5   Sunshine       123710 non-null  float64
 6   WindGustDir    123710 non-null  object 
 7   WindGustSpeed  123710 non-null  float64
 8   WindDir9am     123710 non-null  object 
 9   WindDir3pm     123710 non-null  object 
 10  WindSpeed9am   123710 non-null  float64
 11  WindSpeed3pm   123710 non-null  float64
 12  Humidity9am    123710 non-null  float64
 13  Humidity3pm    123710 non-null  float64
 14  Pressure9am    123710 non-null  float64
 15  Pressure3pm    123710 non-null  float64
 16  Cloud9am       123710 non-null  float64
 17  Cloud3pm       123710 non-null  fl

Berdasarkan informasi di atas, diketahui bahwa data terdiri dari `145460 baris` dan `23 kolom`. Data tersebut masih terdapat missing value terlihat dari 23 kolom, hanya 2 kolom yang berisi data tidak null, yakni kolom Date dan Location. Maka dari itu, diperlukan pembersihan data.

## Data Cleaning

### Examine Duplicated Values

In [10]:
df_wth.duplicated().sum()

0

### Examining Missing Values

In [13]:
df_wth.isnull().sum()

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64

In [29]:
df_wth.drop("Date", axis=1, inplace=True)

### Filling Missing Values
Filling Numerical Columns With Median

In [30]:
numerical = [col for col in df_wth.columns if df_wth[col].dtype!='O']
numerical

['MinTemp',
 'MaxTemp',
 'Rainfall',
 'Evaporation',
 'Sunshine',
 'WindGustSpeed',
 'WindSpeed9am',
 'WindSpeed3pm',
 'Humidity9am',
 'Humidity3pm',
 'Pressure9am',
 'Pressure3pm',
 'Cloud9am',
 'Cloud3pm',
 'Temp9am',
 'Temp3pm']

In [31]:
df_wth.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

In [36]:
df_wth[numerical] = df_wth[numerical].fillna(df_wth[numerical].median())
df_wth.isna().sum()

Location             0
MinTemp              0
MaxTemp              0
Rainfall             0
Evaporation          0
Sunshine             0
WindGustDir      10326
WindGustSpeed        0
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am         0
WindSpeed3pm         0
Humidity9am          0
Humidity3pm          0
Pressure9am          0
Pressure3pm          0
Cloud9am             0
Cloud3pm             0
Temp9am              0
Temp3pm              0
RainToday         3261
RainTomorrow      3267
dtype: int64

### Drop the missing value in Categorical Columns

In [37]:
df_wth.dropna(subset=['WindGustDir','WindDir9am','WindDir3pm','RainToday','RainTomorrow'], axis=0, inplace=True)
df_wth.isna().sum()

Location         0
MinTemp          0
MaxTemp          0
Rainfall         0
Evaporation      0
Sunshine         0
WindGustDir      0
WindGustSpeed    0
WindDir9am       0
WindDir3pm       0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Cloud9am         0
Cloud3pm         0
Temp9am          0
Temp3pm          0
RainToday        0
RainTomorrow     0
dtype: int64

In [39]:
df_wth.shape

(123710, 22)

Setalah dilakukan penghapusan missing values pada categorical value, baris data yang semula `145460` berkurang menjadi `123710` baris.